### ISSUE TITLE
{"error": "disk I/O error: permission denied"}

### ISSUE THREAD

**Issue by mikebryant** [kyverno/policy-reporter#362]
It seems to work fine on startup for a few minutes, then we start getting the following errors

Can see by inspection of the container on the host that the sqlite database on disk is changing, and the journal file appearing and disappearing.

Somewhat confused what could be wrong

```
2023-11-10 15:43:39	ERROR	failed to load data	{"error": "disk I/O error: permission denied"}
github.com/kyverno/policy-reporter/pkg/api/v1.(*Handler).logError
	/app/pkg/api/v1/handler.go:22
github.com/kyverno/policy-reporter/pkg/api.(*httpServer).RegisterV1Handler.(*Handler).NamespacedResourcesKindListHandler.func8
	/app/pkg/api/v1/handler.go:127
github.com/kyverno/policy-reporter/pkg/api.(*httpServer).middleware.Gzip.func1
	/app/pkg/api/gzip.go:48
net/http.HandlerFunc.ServeHTTP
	/usr/local/go/src/net/http/server.go:2136
net/http.(*ServeMux).ServeHTTP
	/usr/local/go/src/net/http/server.go:2514
github.com/kyverno/policy-reporter/pkg/api.NewServer.NewLoggerMiddleware.func1
	/app/pkg/api/server.go:160
net/http.HandlerFunc.ServeHTTP
	/usr/local/go/src/net/http/server.go:2136
net/http.serverHandler.ServeHTTP
	/usr/local/go/src/net/http/server.go:2938
net/http.(*conn).serve
	/usr/local/go/src/net/http/server.go:2009
2023-11-10 15:43:39	ERROR	failed to load namespaced status counts	{"error": "disk I/O error: permission denied"}
github.com/kyverno/policy-reporter/pkg/database.(*Store).FetchNamespacedStatusCounts
	/app/pkg/database/bun.go:575
github.com/kyverno/policy-reporter/pkg/api.(*httpServer).RegisterV1Handler.(*Handler).NamespacedResourcesStatusCountsHandler.func12
	/app/pkg/api/v1/handler.go:198
github.com/kyverno/policy-reporter/pkg/api.(*httpServer).middleware.Gzip.func1
	/app/pkg/api/gzip.go:48
net/http.HandlerFunc.ServeHTTP
	/usr/local/go/src/net/http/server.go:2136
net/http.(*ServeMux).ServeHTTP
	/usr/local/go/src/net/http/server.go:2514
github.com/kyverno/policy-reporter/pkg/api.NewServer.NewLoggerMiddleware.func1
	/app/pkg/api/server.go:160
net/http.HandlerFunc.ServeHTTP
	/usr/local/go/src/net/http/server.go:2136
net/http.serverHandler.ServeHTTP
	/usr/local/go/src/net/http/server.go:2938
net/http.(*conn).serve
	/usr/local/go/src/net/http/server.go:2009
```

```
<99e3-65cd-472d-9675-35ac59e55d5e/volumes/kubernetes.io~empty-dir/sqlite -h
total 45M
drwxrwsrwx. 2 root 1234 4.0K Nov 10 15:46 .
drwxr-xr-x. 3 root root 4.0K Nov 10 15:27 ..
-rw-r--r--. 1 1234 1234  45M Nov 10 15:46 database.db
-rw-r--r--. 1 1234 1234  13K Nov 10 15:46 database.db-journal
<cd-472d-9675-35ac59e55d5e/volumes/kubernetes.io~empty-dir/sqlite -h
total 45M
drwxrwsrwx. 2 root 1234 4.0K Nov 10 15:46 .
drwxr-xr-x. 3 root root 4.0K Nov 10 15:27 ..
-rw-r--r--. 1 1234 1234  45M Nov 10 15:46 database.db
```

We're on helm chart version 2.21.1, EKS v1.25.14-eks-f8587cb

Config is:

```
        global:
          plugins:
            kyverno: true
        kyvernoPlugin:
          enabled: true
        metrics:
          enabled: true
        ui:
          enabled: true
          ingress:
            ... some stuff
```

**Comment by fjogeleit** [comment]
Hey, I’ll have a look on it during next week.

**Comment by fjogeleit** [comment]
Do you have some details about your environment? Do you customize the volume somehow or is it an default emptyDir?

**Comment by mikebryant** [comment]
We're on EKS 1.25 with bottlerocket Nodes (`bottlerocket-aws-k8s-1.25-x86_64-v1.14.3-764e37e4`), using the chart with the above values

The generated volume section is:
```
  volumes:
  - emptyDir: {}
    name: sqlite
```

**Comment by fjogeleit** [comment]
Can you try to add this to your values:

```yaml
sqliteVolume:
    emptyDir:
        medium: Memory
```

**Comment by fjogeleit** [comment]
In general I think that this issue is caused by the provisioning of the emptyDir by your cluster. I don’t find anything I could do on my side.

you can also use any other type of volume configuration instead of an emptyDir and check if this helps.

**Comment by fjogeleit** [comment]
Let me know if you still encounter errors.

**Comment by chdorb** [comment]
Same errors on AKS 1.27.7 with both default and Memory sqlite volume
```
  volumes:
  - emptyDir: {}
    name: sqlite
```
```
  volumes:
  - emptyDir:
      medium: Memory
    name: sqlite
```
```
1.7016845518019803e+09  error   failed to load data     {"error": "disk I/O error: permission denied"}
1.7016845593510585e+09  error   failed to load data     {"error": "disk I/O error: permission denied"}
1.7016845593617969e+09  error   failed to load namespaced status counts {"error": "disk I/O error: permission denied"}
1.70168455936183e+09    error   failed to load data     {"error": "disk I/O error: permission denied"}
1.7016845604148169e+09  error   failed to load data     {"error": "disk I/O error: permission denied"}
1.7016845604278517e+09  error   failed to load data     {"error": "disk I/O error: permission denied"}
1.7016845614776087e+09  error   failed to load data     {"error": "disk I/O error: permission denied"}
1.7016845625346801e+09  error   failed to load data     {"error": "disk I/O error: permission denied"}
1.7016845701141255e+09  error   failed to load namespaced status counts {"error": "disk I/O error: permission denied"}
1.7016845701141646e+09  error   failed to load data     {"error": "disk I/O error: permission denied"}
1.7016845701244414e+09  error   failed to load data     {"error": "disk I/O error: permission denied"}
```

**Comment by chdorb** [comment]
Possible to re-open @fjogeleit ?

**Comment by fjogeleit** [comment]
I am not sure how this should be handled on my side. The issue relates to the bootstrapping of the emptyDir in the underlaying cluster, so depending on your configuration you need to customize the volume settings.

I'll do a second research but I don't think I can do much about it.

**Comment by fjogeleit** [comment]
Maybe worth to mention that policy-reporter does by default not run as root user and has set a fsGroup and runAsUser configuration:

```yaml

podSecurityContext:
  fsGroup: 1234

securityContext:
  runAsUser: 1234
  runAsNonRoot: true
  privileged: false
  allowPrivilegeEscalation: false
  readOnlyRootFilesystem: true
  capabilities:
    drop:
      - ALL
  seccompProfile:
    type: RuntimeDefault
```

maybe this leads to permission issues with your volume and you have to modify this for your setup.

**Comment by mikebryant** [comment]
Sorry - I went on annual leave and then lost track of this

I've just been stracing and found:

```
newfstatat(AT_FDCWD, "/var/tmp", 0x7ffd88ce7d10, 0) = -1 ENOENT (No such file or directory)
newfstatat(AT_FDCWD, "/usr/tmp", 0x7ffd88ce7d10, 0) = -1 ENOENT (No such file or directory)
newfstatat(AT_FDCWD, "/tmp", 0x7ffd88ce7d10, 0) = -1 ENOENT (No such file or directory)
newfstatat(AT_FDCWD, ".", {st_mode=S_IFDIR|0755, st_size=4096, ...}, 0) = 0
access(".", W_OK|X_OK)                  = -1 EACCES (Permission denied)
```
It looks like it's trying to access the current directory (Which I guess is `/app` from the Dockerfile WORKDIR?), which might be failing because of `readOnlyRootFilesystem: true`?

**Comment by mikebryant** [comment]
I have worked around this with:

```
        extraVolumes:
          volumeMounts:
          - mountPath: /tmp
            name: tmp
          volumes:
          - emptyDir: {}
            name: tmp
```

The errors have stopped.

New strace:
```
newfstatat(AT_FDCWD, "/var/tmp", 0x7ffc40039f40, 0) = -1 ENOENT (No such file or directory)
newfstatat(AT_FDCWD, "/usr/tmp", 0x7ffc40039f40, 0) = -1 ENOENT (No such file or directory)
newfstatat(AT_FDCWD, "/tmp", {st_mode=S_IFDIR|S_ISGID|0777, st_size=4096, ...}, 0) = 0
access("/tmp", W_OK|X_OK)               = 0
access("/tmp/etilqs_65bd523eeba405f7", F_OK) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/tmp/etilqs_65bd523eeba405f7", O_RDWR|O_CREAT|O_EXCL|O_NOFOLLOW|O_CLOEXEC, 0600) = 11
newfstatat(11, "", {st_mode=S_IFREG|0600, st_size=0, ...}, AT_EMPTY_PATH) = 0
unlink("/tmp/etilqs_65bd523eeba405f7")  = 0
```

Could this be reopened? Either something needs to stop trying to write to a temporary file, or the chart should be updated with a /tmp emptyDir so it can write files with the default `readOnlyRootFilesystem: true` set

**Comment by fjogeleit** [comment]
@mikebryant @chdorb released with Helm Chart version `2.21.5`

**Comment by chdorb** [comment]
Thank you very much
